primitives/tracing/src/proxy.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::cell::RefCell;
+use rental;
+use tracing::info_span;
+
+/// Used to identify a proxied WASM trace
+pub const WASM_TRACE_IDENTIFIER: &'static str = "WASM_TRACE";
+/// Used to extract the real `target` from the associated values of the span
+pub const WASM_TARGET_KEY: &'static str = "proxied_wasm_target";
+/// Used to extract the real `name` from the associated values of the span
+pub const WASM_NAME_KEY: &'static str = "proxied_wasm_name";
+
+const MAX_SPANS_LEN: usize = 1000;
+
+thread_local! {
+	static PROXY: RefCell<TracingProxy> = RefCell::new(TracingProxy::new());
```

Comment:
Why do we need a thread local here? Because we need to store the stuff somewhere?

You should register dynamically an extension in the runtime interface. Search for `register_extension` in `sp-io` for examples.

If we do this by an extension, we make sure that everything is dropped when we leave the wasm runtime. As this "leaving" could also be a panic which would lead to these spans reside in here until the end of the program (not good).

Reply:
Ah ok, I hadn't realised that panic would mean these aren't dropped - so the way you suggest definitely makes sense.